client/db/src/lib.rs

```diff
@@ -1560,6 +1583,53 @@ fn apply_state_commit(transaction: &mut Transaction<DbHash>, commit: sc_state_db
 	}
 }
 
+fn apply_index_ops<Block: BlockT>(
+	transaction: &mut Transaction<DbHash>,
+	body: Vec<Block::Extrinsic>,
+	ops: Vec<IndexOperation>
+) -> Vec<u8> {
+	let mut extrinsic_headers = Vec::with_capacity(body.len());
+	let mut index_map = HashMap::new();
+	let mut renewed_map = HashMap::new();
+	for op in ops {
+		match op {
+			IndexOperation::Insert { extrinsic, offset } => {
+				index_map.insert(extrinsic, offset);
+			}
+			IndexOperation::Renew { extrinsic, hash, .. } => {
+				renewed_map.insert(extrinsic, DbHash::from_slice(hash.as_ref()));
+			}
+		}
+	}
+	for (index, extrinsic) in body.into_iter().enumerate() {
+		let extrinsic = extrinsic.encode();
+		if let Some(hash) = renewed_map.get(&(index as u32)) {
+			// Bump ref counter
+			transaction.store(columns::TRANSACTION, DbHash::from_slice(hash.as_ref()), None);
+			extrinsic_headers.push((hash.clone(), extrinsic));
+			continue;
+		}
+		let offset = match index_map.get(&(index as u32)) {
+			Some(offset) if *offset as usize <= extrinsic.len() => *offset as usize,
+			_ => extrinsic.len(),
+		};
+		let hash = HashFor::<Block>::hash(&extrinsic[offset..]);
+		transaction.store(
+			columns::TRANSACTION,
+			DbHash::from_slice(hash.as_ref()),
+			Some(extrinsic[offset..].to_vec())
+		);
+		extrinsic_headers.push((DbHash::from_slice(&hash.as_ref()), extrinsic[..offset].to_vec()));
```

Comment:
> extrinsic_headers is mixing extrinsic using applied offset with full extrinsic from renewed, is it ok?

Yes, all extrinsics that don't index anything, including renewals, are just written as header.

> Also wondering how/if 'extrinsic[..offset]' can be decoded?

It is never decoded on its own. Only full extrinsic (header + indexed portion) is decoded. The indexed region is not supposed to be scale-encoded. It's just some bytes that are interpreted by the user.

> Generally I am not sure how 'renewed' is supposed to use its 'size' field.

`size` field is a bit of work in progress. The idea is that the renewal transaction should include fee that is proportional to the existing data size. Here, in commit we would check existing expected indexed data size and don't renew anything if it does not match. However it is not clear yet if actually checking the size in the runtime would be better.

Reply:
(for whatever reason I was thinking renewed did contain the indexed content, and did need to be strip, but obviously it doesn't).
I do not mark the conversation as resolved because the reply looks useful to me, but it is resolved.